File: asynchronous/src/main/java/com/dao/asynchronous/OnAsynchronousListener.java
package com.dao.asynchronous;

/**
 * Created in 08/11/18 11:00.
 *
 * @author Diogo Oliveira.
 */
public abstract class OnAsynchronousListener<Result> implements OnCompleteAsynchronousListener<Result>
{
    @Override
    public void onBegin(String tag) { }

    @Override
    public void onSuccess(String tag, Result result) { }

    @Override
    public void onProgress(String tag, Progress progress) { }

    @Override
    public void onCancelled(String tag, Result result) { }

    @Override
    public void onError(String tag, Exception e) { }

    @Override
    public void onFinish(Asynchronous task) { }
}

File: asynchronous/src/main/java/com/dao/asynchronous/OnAsynchronousListener.kt
package com.dao.asynchronous

/**
 * Created in 30/01/19 19:36.
 *
 * @author Diogo Oliveira.
 */
interface OnAsynchronousListener<Result>
{
    fun onBegin(tag: String) {}

    fun onSuccess(tag: String, result: Result?)

    fun onProgress(tag: String, progress: Progress) {}

    fun onCancelled(tag: String, result: Result?) {}

    fun onError(tag: String, e: Exception) {}

    fun onFinish(task: AsyncFinished) {}
}